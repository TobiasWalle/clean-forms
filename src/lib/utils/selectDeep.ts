import { assertPropertyInObject } from './assertPropertyInObject';

export function selectDeep(object: any, path: string[]): any {
  return path.reduce((item, key: string) => {
    assertPropertyInObject(item, key);
    return item[key];
  }, object);
}
