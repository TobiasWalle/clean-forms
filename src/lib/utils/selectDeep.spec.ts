import { selectDeep } from './selectDeep';

describe('selectDeep', () => {
  it('should select an property of an object by the path', () => {
    const object = { a: { b: { c: 123 } } };
    const path = ['a', 'b', 'c'];

    const result = selectDeep(object, path);

    expect(result).toBe(123);
  });

  it('should select an property of an object by the path with a length of 1', () => {
    const object = { a: { b: { c: 123 } } };
    const path = ['a'];

    const result = selectDeep(object, path);

    expect(result).toEqual({ b: { c: 123 } });
  });

  it('should return the object if the path has a length of 0', () => {
    const object = { a: { b: { c: 123 } } };
    const path: string[] = [];

    const result = selectDeep(object, path);

    expect(result).toBe(object);
  });

  it('should throw an error if the path does not exist', () => {
    const object = { a: { b: { c: 123 } } };
    const path = ['b'];

    expect(() => selectDeep(object, path)).toThrowErrorMatchingSnapshot();
  });
});
