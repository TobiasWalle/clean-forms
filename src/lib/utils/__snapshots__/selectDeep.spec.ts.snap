// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectDeep should throw an error if the path does not exist 1`] = `"The key \\"b\\" does not exits on item {\\"a\\":{\\"b\\":{\\"c\\":123}}}."`;
